File: java/0853-car-fleet/Solution.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.Stack;

public class Solution {
    public int carFleet(int target, int[] position, int[] speed) {
        int length = position.length;

        if (length == 1) {
            return 1;
        }

        int[][] cars = new int[length][2];
        for (int i = 0; i < length; ++i) {
            cars[i][0] = position[i];
            cars[i][1] = speed[i];
        }
        Arrays.sort(cars, Comparator.comparingInt(i -> i[0]));

        Stack<Double> stack = new Stack<>();
        for (int i = length - 1; i >= 0; --i) {
            double time = (double) (target - cars[i][0]) / cars[i][1];

            if (!stack.isEmpty() && time <= stack.peek()) {
                continue;
            } else {
                stack.push(time);
            }
        }

        return stack.size();
    }
}

File: c++/0853-car-fleet/solution.cpp
#include <algorithm>
#include <utility>
#include <vector>

using namespace std;

class Solution {
public:
    int carFleet(int target, vector<int>& position, vector<int>& speed) {
        int length = position.size();

        if (length == 1) {
            return 1;
        }

        vector<pair<int, double>> cars;
        for (int i = 0; i < length; ++i) {
            double time = (double) (target - position[i]) / speed[i];

            cars.push_back({position[i], time});
        }
        sort(cars.begin(), cars.end());

        double maxTime = 0.0;
        int result = 0;

        for (int i = length - 1; i >= 0; --i) {
            if (cars[i].second > maxTime) {
                maxTime = cars[i].second;
                ++result;
            }
        }

        return result;
    }
};
